File: backend/app/core/capabilities.py
"""Edition capabilities / entitlements (cloud-edition seam).

Server-driven feature gating: the backend decides which feature surfaces
exist and the frontend renders only what ``GET /system/capabilities``
returns.

Community/self-hosted default: **everything ON** (zero behavior change).
The commercial cloud edition replaces the resolver via
``set_capability_resolver`` to compute ``edition ∩ subscription tier``
(e.g. hide engine tuning and watch folders from tenants, gate BYOK-LLM to a
paid tier). Capability KEYS are generic core vocabulary; the cloud
*resolution policy* is proprietary and lives in the private layer.

Rule: every gate — backend dependency, beat schedule, frontend tab — reads
the same capability key. Never sprinkle ``if edition == "cloud"`` in feature
code.

What a capability actually enforces today
-----------------------------------------
The keys below are declared in one map but are **not** all enforced to the
same depth. Read this before assuming a key protects anything:

1. **Server-enforced** — wired to a ``require_capability`` router dependency
   in ``app/api/router.py``; disabling one makes every route under that
   router 404, so the UI gate is genuinely cosmetic:
   ``watch_sources``, ``organizations``, ``auth.config_ui``,
   ``llm.user_settings``, ``asr.user_providers``, ``engine.settings``.

2. **UI-only vocabulary** — read *only* by ``SettingsModal.svelte`` to hide a
   settings panel. The underlying endpoints stay fully reachable, so this is
   presentation, not authorization: ``recording``, ``exports``,
   ``transcription.prefs``, ``vocab.user``, ``prompts.user``,
   ``redaction.user``, ``sharing.teams``, ``redaction.policy``, ``billing``,
   ``usage_dashboard``, ``users.local_admin``, ``system.hardware_stats``,
   ``audit.logs``, and the ``admin.*`` platform panels.

3. **Declared but unread** — no router, no task, no UI reads these yet.
   Setting ``upload: False`` or ``search: False`` changes nothing at all:
   ``upload``, ``url_ingest``, ``search``, ``comments``,
   ``collections.shared``, ``speakers.shared``, ``prompts.shared``,
   ``org.defaults``, ``asr.model_selection``, ``llm.byok``.

Tiers 2 and 3 are a to-do list, not a contract. A key becomes authoritative
only when a router (or task/beat) reads it — attaching one is a behavior
change and needs its own review.

``tests/unit/test_capability_contract.py`` pins the frontend's ``cap:``
strings as a subset of ``COMMUNITY_CAPABILITIES``: the two maps previously
drifted in opposite directions, and because unknown keys fail **closed**
here but **open** in the frontend store, the drift was invisible until a key
was finally declared.
"""

import logging
from collections.abc import Callable

from fastapi import HTTPException
from fastapi import Request

from app.core.config import settings

logger = logging.getLogger(__name__)

# ---- Audience taxonomy -------------------------------------------------------
# WHO a surface is for when it is enabled. Orthogonal to enabled-ness: the
# capability map says whether the surface EXISTS in this deployment; the
# audience says which role's UI/permissions it belongs to. End users are
# never admins; org admins manage their tenant only (billing, members,
# policies) and are NEVER platform admins; "platform" = Attevon staff in
# cloud / the operator in self-host.
AUDIENCE_USER = "user"  # individual end users (their own content/prefs)
AUDIENCE_TEAM = "team"  # inner-team collaboration among org/group members
AUDIENCE_ORG_ADMIN = "org_admin"  # tenant administrators (billing, seats, policies)
AUDIENCE_PLATFORM = "platform"  # platform staff / self-host operator (config & ops)

# Known capability keys with their COMMUNITY defaults (everything on except
# the cloud-only surfaces, which don't exist without the cloud layer).
COMMUNITY_CAPABILITIES: dict[str, bool] = {
    # -- user-facing ------------------------------------------------------------
    "upload": True,  # file upload + presigned flow
    "recording": True,  # in-browser recording
    "url_ingest": True,  # yt-dlp URL ingestion
    "search": True,  # hybrid/semantic search
    "exports": True,  # SRT/VTT/TXT/bulk exports
    "comments": True,  # per-transcript comments
    "transcription.prefs": True,  # language/translate/speaker prefs
    "vocab.user": True,  # custom vocabulary (own terms)
    "prompts.user": True,  # own summary prompts
    "redaction.user": True,  # personal redaction preferences
    "asr.user_providers": True,  # per-user cloud-ASR provider configs + keys
    "llm.user_settings": True,  # per-user LLM provider/endpoint/keys
    "watch_sources": True,  # local/S3/SMB auto-import (Settings → Watch Sources)
    # -- team-facing (inner-team collaboration) ----------------------------------
    "sharing.teams": True,  # groups + collection shares
    "collections.shared": True,  # shared/org-visible collections
    "speakers.shared": True,  # shared speaker profiles
    "prompts.shared": True,  # shared prompt library
    # -- org-admin (tenant administration; cloud-only surfaces) ------------------
    "billing": False,
    "usage_dashboard": False,
    "organizations": False,  # member/seat management surface
    "org.defaults": False,  # org-level default settings/policies
    "llm.byok": False,  # bring-your-own LLM key (cloud tier-gated extra)
    "redaction.policy": True,  # enforcement floor (org admin in cloud; operator here)
    # -- platform (staff / self-host operator config & ops) ----------------------
    "auth.config_ui": True,  # LDAP/Keycloak/PKI admin configuration UI
    "users.local_admin": True,  # local user CRUD admin UI
    "asr.model_selection": True,  # admin local-model pinning UI
    "engine.settings": True,  # diarization/boundary tuning admin panel
    "system.hardware_stats": True,  # GPU/CPU stats, benchmarks
    "audit.logs": True,  # audit log viewer (super-admin)
    "admin.data_integrity": True,  # orphaned-record / storage reconciliation panel
    "admin.retention": True,  # file retention + derived-media cache policy
    "admin.backup": True,  # scheduled database backups
    "admin.search_indexing": True,  # reindex / embedding-model management
    "admin.embedding_migration": True,  # embedding migration + consistency panel
    "admin.task_health": True,  # stuck-task recovery / queue health panel
}

# Every capability key MUST be classified, and the two maps must carry the
# SAME key set (both pinned by tests). The frontend uses this to place
# surfaces in the right area (user settings / team / org admin / platform
# admin) — and to render nothing at all for roles below the audience.
CAPABILITY_AUDIENCE: dict[str, str] = {
    "upload": AUDIENCE_USER,
    "recording": AUDIENCE_USER,
    "url_ingest": AUDIENCE_USER,
    "search": AUDIENCE_USER,
    "exports": AUDIENCE_USER,
    "comments": AUDIENCE_USER,
    "transcription.prefs": AUDIENCE_USER,
    "vocab.user": AUDIENCE_USER,
    "prompts.user": AUDIENCE_USER,
    "redaction.user": AUDIENCE_USER,
    "asr.user_providers": AUDIENCE_USER,
    "llm.user_settings": AUDIENCE_USER,
    "watch_sources": AUDIENCE_USER,
    "sharing.teams": AUDIENCE_TEAM,
    "collections.shared": AUDIENCE_TEAM,
    "speakers.shared": AUDIENCE_TEAM,
    "prompts.shared": AUDIENCE_TEAM,
    "billing": AUDIENCE_ORG_ADMIN,
    "usage_dashboard": AUDIENCE_ORG_ADMIN,
    "organizations": AUDIENCE_ORG_ADMIN,
    "org.defaults": AUDIENCE_ORG_ADMIN,
    "llm.byok": AUDIENCE_ORG_ADMIN,
    "redaction.policy": AUDIENCE_ORG_ADMIN,
    "auth.config_ui": AUDIENCE_PLATFORM,
    "users.local_admin": AUDIENCE_PLATFORM,
    "asr.model_selection": AUDIENCE_PLATFORM,
    "engine.settings": AUDIENCE_PLATFORM,
    "system.hardware_stats": AUDIENCE_PLATFORM,
    "audit.logs": AUDIENCE_PLATFORM,
    "admin.data_integrity": AUDIENCE_PLATFORM,
    "admin.retention": AUDIENCE_PLATFORM,
    "admin.backup": AUDIENCE_PLATFORM,
    "admin.search_indexing": AUDIENCE_PLATFORM,
    "admin.embedding_migration": AUDIENCE_PLATFORM,
    "admin.task_health": AUDIENCE_PLATFORM,
}

# Resolver signature: (request | None) -> capability dict. The request is
# offered so a cloud resolver can derive tenant/tier from the verified
# identity stashed by get_current_user (request.state.external_identity).
CapabilityResolver = Callable[[Request | None], dict[str, bool]]


def _community_resolver(_request: Request | None) -> dict[str, bool]:
    return dict(COMMUNITY_CAPABILITIES)


_resolver: CapabilityResolver = _community_resolver


def set_capability_resolver(resolver: CapabilityResolver) -> None:
    """Replace the capability resolver (registered by the cloud layer)."""
    global _resolver
    logger.info("Capability resolver overridden (cloud edition)")
    _resolver = resolver


def reset_capability_resolver() -> None:
    """Restore the community resolver (primarily for tests)."""
    global _resolver
    _resolver = _community_resolver


def get_capabilities(request: Request | None = None) -> dict[str, bool]:
    """Effective capability map for this deployment/request.

    Unknown keys from a custom resolver are passed through; missing known
    keys fall back to the community defaults so a partial resolver cannot
    accidentally disable surfaces it never considered.
    """
    resolved = _resolver(request)
    caps = dict(COMMUNITY_CAPABILITIES)
    caps.update(resolved)
    return caps


def capability_enabled(key: str, request: Request | None = None) -> bool:
    """Check a single capability (False for unknown keys)."""
    return bool(get_capabilities(request).get(key, False))


def require_capability(key: str, *, platform_admin_bypass: bool = True) -> Callable:
    """FastAPI dependency factory: 404 when the capability is off.

    404 (not 403) on purpose — a disabled feature surface should not exist
    at all for this deployment, exactly like an unknown route.

    ``platform_admin_bypass`` (default on) lets platform staff
    (``is_superuser``) through even when the capability is hidden from
    tenants — cloud tenants can never become superusers (external_sync never
    grants it), so this exactly implements the capabilities matrix's
    "invisible to org users / available to platform staff" column.
    """

    def _dependency(request: Request) -> None:
        if capability_enabled(key, request):
            return
        if platform_admin_bypass:
            try:
                from app.api.endpoints.auth import get_optional_current_user
                from app.db.base import SessionLocal

                db = SessionLocal()
                try:
                    user = get_optional_current_user(request, db)
                    if user is not None and bool(user.is_superuser):
                        return
                finally:
                    db.close()
            except Exception:
                # Bypass is best-effort; any failure falls through to 404.
                logger.debug("platform-admin bypass check failed", exc_info=True)
        raise HTTPException(status_code=404, detail="Not Found")

    return _dependency


def edition() -> str:
    """Deployment edition string ("community" | "cloud")."""
    return settings.DEPLOYMENT_EDITION

File: frontend/src/stores/capabilities.ts
/**
 * Edition capabilities / entitlements store.
 *
 * Mirrors GET /api/system/capabilities — the backend decides which feature
 * surfaces exist for this deployment (community: everything; cloud: the
 * managed subset, tier-aware). The UI renders only enabled surfaces; the
 * backend independently 404s disabled endpoints, so this is purely cosmetic
 * gating on top of a real server-side gate.
 *
 * Fail-open by design: until the endpoint responds (or if it ever fails),
 * the store holds community defaults so a self-hosted instance — or a blip
 * during startup — never hides features.
 */

import { writable } from 'svelte/store';
import axiosInstance from '$lib/axios';

type CapabilityAudience = 'user' | 'team' | 'org_admin' | 'platform';

export interface CapabilitiesState {
  edition: 'community' | 'cloud';
  loaded: boolean;
  capabilities: Record<string, boolean>;
  audience: Record<string, CapabilityAudience>;
  /**
   * The live, admin-configurable upload ceiling (`settings.MAX_UPLOAD_BYTES`) from a
   * successful `/system/capabilities` fetch: a byte count, or `null` when the admin
   * explicitly disabled it (`MAX_UPLOAD_BYTES=0`). `undefined` until we actually know —
   * before the first fetch resolves, or if it failed. Deliberately NOT defaulted to a
   * number here: `$lib/utils/uploadLimits` (the owner of the upload-limit fallback
   * constant) treats `undefined` as "fall back to the coded default", so a fetch that
   * hasn't resolved yet never reads as "no limit".
   */
  maxUploadBytes: number | null | undefined;
}

const COMMUNITY_DEFAULTS: CapabilitiesState = {
  edition: 'community',
  loaded: false,
  capabilities: {},
  audience: {},
  maxUploadBytes: undefined,
};

export const capabilities = writable<CapabilitiesState>(COMMUNITY_DEFAULTS);

/**
 * True unless the backend explicitly disabled the key. Unknown keys default
 * to enabled (fail-open) — the cloud resolver explicitly sets false for
 * every surface it hides.
 *
 * This is the OPPOSITE of the backend's `capability_enabled`, which returns
 * false for an unknown key (fail-closed). The asymmetry is intentional (a
 * blank/failed capabilities fetch must not blank the self-hosted UI) but it
 * hides drift: a `cap:` key the backend has never declared renders here
 * forever and vanishes the moment someone declares it as `False`. Hence
 * `backend/tests/unit/test_capability_contract.py`, which pins every `cap:`
 * string in SettingsModal.svelte to a declared backend capability.
 */
export function isCapabilityEnabled(state: CapabilitiesState, key: string): boolean {
  return state.capabilities[key] !== false;
}

/**
 * Drop the fetched capability map back to community defaults (`loaded: false`).
 *
 * Registered in `$lib/session/clearUserState`. Without it, the cloud edition
 * leaked one user's TIER-SCOPED capability map into the next session on the same
 * tab: `loadCapabilities()` has exactly one call site (`routes/+layout.svelte`
 * `onMount`), and an SPA login does not re-run `onMount`, so User B saw User A's
 * enabled surfaces until a hard reload.
 *
 * Resetting to `loaded: false` (not `loaded: true`) is deliberate: consumers that
 * wait for `loaded` must wait for the NEXT user's fetch, not read the fail-open
 * empty map as an authoritative answer.
 */
export function resetCapabilities(): void {
  capabilities.set({ ...COMMUNITY_DEFAULTS, capabilities: {}, audience: {} });
}

/** Fetch once at app bootstrap; safe to call again (e.g. after login). */
export async function loadCapabilities(): Promise<void> {
  try {
    const response = await axiosInstance.get('/system/capabilities');
    const rawMaxUpload: unknown = response.data?.max_upload_bytes;
    capabilities.set({
      edition: response.data?.edition === 'cloud' ? 'cloud' : 'community',
      loaded: true,
      capabilities: response.data?.capabilities ?? {},
      audience: response.data?.audience ?? {},
      // Explicit `null` (admin disabled the limit) must survive as `null`, not fall
      // into the "unknown" bucket — only an actually missing/malformed field does.
      maxUploadBytes:
        rawMaxUpload === null ? null : typeof rawMaxUpload === 'number' ? rawMaxUpload : undefined,
    });
  } catch {
    // Fail-open for feature visibility (community defaults, everything shown), but NOT
    // for the upload ceiling: `maxUploadBytes` stays `undefined` so
    // `$lib/utils/uploadLimits` falls back to its own coded default rather than reading
    // a failed fetch as "no limit".
    capabilities.set({ ...COMMUNITY_DEFAULTS, loaded: true });
  }
}
